ADAL/src/workplacejoin/mac/ADWorkPlaceJoinUtil.m

```diff
@@ -22,14 +22,314 @@
+                                     error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
+{
+    OSStatus status= noErr;
+    SecCertificateRef certRef = NULL;
+    NSData *issuerTag = [self wpjCertIssuerTag];
+    
+    // Set the private key query dictionary.
+    NSDictionary *queryCert = @{ (__bridge id)kSecClass : (__bridge id)kSecClassCertificate,
+                                 (__bridge id)kSecAttrLabel : issuerTag
+                                 };
+    
+    // Get the certificate. If the certificate is not found, this is not considered an error.
+    status = SecItemCopyMatching((__bridge CFDictionaryRef)queryCert, (CFTypeRef*)&certRef);
+    if (status == errSecItemNotFound)
+    {
+        return NULL;
+    }
+    
+    CHECK_KEYCHAIN_STATUS(@"Failed to read WPJ certificate.");
+    
+    return certRef;
+    
+_error:
+    return NULL;
+}
+
++ (NSData *)wpjCertIssuerTag
+{
+    return [NSData dataWithBytes:certificateIdentifier length:strlen((const char *)certificateIdentifier)];
+}
+
++ (SecKeyRef)copyPrivateKeyRefForIdentifier:(NSString *)identifier
+                                    context:(id<ADRequestContext>)context
+                                      error:(ADAuthenticationError* __nullable __autoreleasing * __nullable)error
+{
+    OSStatus status= noErr;
+    SecKeyRef privateKeyReference = NULL;
+    
+    NSData *privateKeyTag = [NSData dataWithBytes:[identifier UTF8String] length:identifier.length];
+    
+    // Set the private key query dictionary.
+    NSDictionary *privateKeyQuery = @{ (__bridge id)kSecClass : (__bridge id)kSecClassKey,
+                                       (__bridge id)kSecAttrApplicationTag : privateKeyTag,
+                                       (__bridge id)kSecAttrKeyType : (__bridge id)kSecAttrKeyTypeRSA,
+                                       (__bridge id)kSecReturnRef : (__bridge id)kCFBooleanTrue
+                                       };
+    
+    // Get the key.
+    status = SecItemCopyMatching((__bridge CFDictionaryRef)privateKeyQuery, (CFTypeRef*)&privateKeyReference);
+    CHECK_KEYCHAIN_STATUS(@"Failed to read WPJ private key for identifier.");
+    
+    return privateKeyReference;
+    
+_error:
+    return nil;
+}
+
++ (nullable NSString *)wpjStringForIdentifier:(nonnull NSString *)identifier
+                                      context:(id<ADRequestContext>)context
+                                        error:(ADAuthenticationError * __nullable __autoreleasing * __nullable)error
 {
-    (void)correlationId;
-    (void)error;
+    // Building dictionary to retrieve UPN from the keychain
+    NSDictionary *query = @{ (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
+                             (__bridge id)kSecAttrAccount : identifier,
+                             (__bridge id)kSecReturnAttributes : (__bridge id)kCFBooleanTrue
+                             };
```

Comment:
A kSecAttrService should be provided here as well reduce the likelihood of collisions in the keychain